torchtune/models/gemma/_component_builders.py

```diff
@@ -121,3 +124,239 @@ def gemma_mlp(dim: int, hidden_dim: int) -> FeedForward:
+            Default: False
+        apply_lora_to_output (bool): whether to apply LoRA to the model's final output projection.
+            Default: False
+        vocab_size (int): number of tokens in vocabulary.
+        num_layers (int): number of layers in the transformer decoder.
+        num_heads (int): number of query heads. For MHA this is also the
+            number of heads for key and value
+        head_dim (int): dimension of head
+        num_kv_heads (int): number of key and value heads.
+        embed_dim (int): embedding dimension for self-attention
+        intermediate_dim (int): intermediate dimension for MLP
+        max_seq_len (int): maximum sequence length the model will be run with,
+        attn_dropout (float): dropout value passed onto scaled_dot_product_attention.
+            Default: 0.0
+        norm_eps (float): epsilon in RMS norms Default: 1e-6
+        rope_base (int): base for the rotary positional embeddings. Default: 10_000
+        norm_embeddings (bool): whether to apply layer norm before the self-attention
+            and mlp layers. Default: True
+        lora_rank (int): rank of each low-rank approximation
+        lora_alpha (float): scaling factor for the low-rank approximation
+        lora_dropout (float): LoRA dropout probability. Default: 0.0
+        quantize_base: (bool): Whether to quantize base model weights or not. Only applied to base
+            weights within linear layers LoRA is applied to. The final output linear projection is not
+            supported for quantization currently.
+
+    Returns:
+        GemmaTransformerDecoder: Instantiation of Gemma model with LoRA applied to
+        a subset of the attention projections in each layer.
+    """
+    self_attn = lora_gemma_self_attention(
+        lora_modules=lora_attn_modules,
+        embed_dim=embed_dim,
+        head_dim=head_dim,
+        num_heads=num_heads,
+        num_kv_heads=num_kv_heads,
+        max_seq_len=max_seq_len,
+        attn_dropout=attn_dropout,
+        rope_base=rope_base,
+        lora_rank=lora_rank,
+        lora_alpha=lora_alpha,
+        lora_dropout=lora_dropout,
+        quantize_base=quantize_base,
+    )
+
+    if apply_lora_to_mlp:
+        mlp = lora_gemma_mlp(
+            dim=embed_dim,
+            hidden_dim=intermediate_dim,
+            lora_rank=lora_rank,
+            lora_alpha=lora_alpha,
+            quantize_base=quantize_base,
+        )
+    else:
+        mlp = gemma_mlp(dim=embed_dim, hidden_dim=intermediate_dim)
+
+    layer = TransformerDecoderLayer(
+        attn=self_attn,
+        mlp=mlp,
+        sa_norm=GemmaRMSNorm(embed_dim, eps=norm_eps),
+        mlp_norm=GemmaRMSNorm(embed_dim, eps=norm_eps),
+    )
+    tok_embeddings = nn.Embedding(vocab_size, embed_dim)
+
+    # TODO: Fix the apply to output_proj.
+    assert apply_lora_to_output is False, "LoRA is not currently supported for the output projection."
```

Comment:
Personally I think this is the right way to handle this (assuming we don't want to support LoRA on the tied embedding weights). Maybe we can also consider not exposing the arg `apply_lora_to_output` at all, since it is basically trivial at this point? 

Re LoRA state dict validation, we could consider just doing e.g. `getattr(cfg_model, "apply_lora_to_output", False)`  instead of our current usage of `cfg_model.apply_lora_to_output`. But cc @RdoubleA for thoughts on the config experience here

Reply:
let's just remove this field entirely, and make a note in the docstring that output projection lora is not supported because it is tied to token embeddings? seems like the simplest approach

Reply:
Thanks, I add `cfg_model.apply_lora_to_output = getattr(cfg_model, "apply_lora_to_output", False)` in `lora_finetune_single_device.py` and `lora_finetune_distributed.py`, and remove the arg `apply_lora_to_output ` in LoRA and QLoRA yaml file.

Reply:
could we also remove it from this builder to prevent confusion?

Reply:
Yes, I remove it.